### ISSUE TITLE
Recommended parameters for transformer model

### ISSUE THREAD

**Issue by franckbrl** [awslabs/sockeye#329]
Hello!

Apologies if this is redundant, but I thought #154 could use a small update, since some things have changed since then.

Here are the arguments I used to train a transformer model on about 25M English-French parallel sentences:

```
python3 -m sockeye.train \
	-s $DATA/my_train.en.bpe \
	-t $DATA/my_train.fr.bpe \
	-vs $DATA/my_dev.en.bpe \
	-vt $DATA/my_dev.fr.bpe \
	--decode-and-evaluate-use-cpu \
	--batch-size 15000 \
	--batch-type word \
	--checkpoint-frequency 10000 \
	--optimized-metric bleu \
	--max-num-checkpoint-not-improved 10 \
	--num-embed 512 \
	--max-seq-len 300 \
	--encoder transformer \
	--decoder transformer \
	--layer-normalization \
	--transformer-model-size 512 \
	--transformer-feed-forward-num-hidden 2048 \
	--transformer-dropout-prepost 0.1 \
	--weight-tying-type src_trg_softmax \
	--weight-tying \
	--initial-learning-rate 0.0003 \
	--learning-rate-warmup 50000 \
	--loss cross-entropy \
	--label-smoothing 0.2 \
	--output my_model
```

With this, I obtained similar results to a Tensor2Tensor model (roughly the same setup). What else would you recommend? No gradient clipping? Layer/weight normalization?

How about adding dropout to other layers (embedding, attention, act)? And values to set --transformer-{pre|post}process?

**Comment by mjpost** [comment]
Did you see [the settings we used](https://github.com/awslabs/sockeye/blob/arxiv_1217/arxiv/code/transformer/sockeye/train-transformer.sh) for the [Sockeye paper](https://arxiv.org/abs/1712.05690)? That disabled gradient clipping (-1), and used "n" (layer normalization) for transformer preprocess and "dr" (dropout and residual connections) for postprocess. (I believe that your `--layer-normalization` has no effect for the transformer model)

**Comment by franckbrl** [comment]
I had only seen the paper. Thank you!

**Comment by mjpost** [comment]
I just [came across this](http://ufallab.ms.mff.cuni.cz/%7Epopel/training-tips-transformer.pdf), too, which might be of use (though this is for tensor2tensor, not sockeye).

**Comment by franckbrl** [comment]
Yes, it's a good one, and I enjoyed the conclusion:  "In sum, experiments done for this article took about 4 years of GPU time."

However, this exploration seems less focused on architecture details than training parameters (batch size, learning rate...). And these are probably very dependent on what they call "learning rate" for Adam optimization ([section 5.3](https://arxiv.org/pdf/1706.03762.pdf)), which is different from what we have in Sockeye.
